### PR TITLE
#19862 Fix division-by zero on low rates

### PR DESCRIPTION
**Pull request by drewhk**
- Fix throttle to start with a full bucket (first commit)
- Fix throttle to support a wide range of rates (see https://github.com/akka/akka/pull/19869 for the earlier approach that used only multiplication but was less flexible)

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/Throttle.scala [hunk 1]
```diff
@@ -20,21 +35,32 @@ private[stream] class Throttle[T](cost: Int,
...
+    var recordedTokens: Long = maximumBurst
+    // Tracks when the last registered token arrived, i.e. the ideal time when the *tracked* bucket jumped
+    // from n to n+1 tokens (but not when it was actually calculated).
+    // This needs to be only updated for token *additions* as subtractions have nothing to do with token arrival.
+    // If 5 tokens arrived while 4 were consumed, this needs to be still increased by 5 * nanosBetweenTokens.
+    // Be aware that this timestamp is usually in the past compared to when the ideal bucket had a new token arrival.
+    // We track here only the timestamp of the last *registered* token, i.e. corresponding to the last addition to
+    // recordedTokens (this is due to our on-need, lazy bucket update policy which makes us always slightly be in
+    // the past compared to the ideal bucket, doing large jumps in the future only when update is needed).
+    var arrivalOfLastRecordedToken: Long = System.nanoTime()
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: this could be 0 actually, since we update it in preStart() (which is slightly closer in time to the "real" start).

File: akka-stream/src/main/scala/akka/stream/impl/Throttle.scala [hunk 2]
```diff
@@ -20,21 +35,32 @@ private[stream] class Throttle[T](cost: Int,
...
   require(cost > 0, "cost must be > 0")
-  require(per.toMillis > 0, "per time must be > 0")
+  require(per.toNanos > 0, "per time must be > 0")
   require(!(mode == ThrottleMode.Enforcing && maximumBurst < 0), "maximumBurst must be > 0 in Enforcing mode")
+  require(per.toNanos >= cost, "Rates larger than 1 unit / nanosecond are not supported")
+
+  // There is some loss of precision here because of rounding, but this only happens if nanosBetweenTokens is very
+  // small which is usually at rates where that precision is highly unlikely anyway as the overhead of this stage
+  // is likely higher than the required accuracy interval.
+  private val nanosBetweenTokens = (per.toNanos.toDouble / cost).toLong
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: I don’t see how `.toDouble` would have an effect here (apart from introducing rounding errors for weirdly specified rates).

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: yeah, right, refactor blindness here :)

File: akka-stream/src/main/scala/akka/stream/impl/Throttle.scala [hunk 3]
```diff
@@ -79,11 +113,8 @@ private[stream] class Throttle[T](cost: Int,
       override def onPull(): Unit = pull(in)
     })
 
-    override def preStart(): Unit = previousTime = now()
-
-    private def now(): Long = System.nanoTime()
+    override def preStart(): Unit = arrivalOfLastRecordedToken = System.nanoTime()
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: this duplicates the initialization above, I think one is enough

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: I noticed already :). It is actually better to do here.

File: akka-stream/src/main/scala/akka/stream/impl/Throttle.scala [hunk 4]
```diff
@@ -42,26 +68,34 @@ private[stream] class Throttle[T](cost: Int,
...
+                arrivalOfLastRecordedToken += elementCost * nanosBetweenTokens
+                scheduleOnce(timerName, waitNanos.nanos)
               case Enforcing ⇒ failStage(new RateExceededException("Maximum throttle throughput exceeded"))
             }
           else {
-            lastTokens = currentTokens - elementCost
-            previousTime = currentTime
+            recordedTokens = refreshedTokens - elementCost
+            // Never set this timestamp to currentNanos directly! This tracks the *ideal* time of the arrival of
+            // the last recorded token, not when we actually calculated it (currentNanos).
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: I think this is incorrect: this only holds unless the bucket was completely full, at which point we cannot fill it further (not even ideally), so we must set the last recorded time to “now”. With the current solution we effectively switch off the burst limit.

File: akka-stream/src/main/scala/akka/stream/impl/Throttle.scala [hunk 5]
```diff
@@ -71,7 +105,7 @@ private[stream] class Throttle[T](cost: Int,
     override protected def onTimer(key: Any): Unit = {
       push(out, currentElement.get)
       currentElement = None
-      lastTokens = 0
+      recordedTokens = 0 // This will be recalculated automatically once the next onPush is invoked
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: This should be moved upwards, next to the modification of the last recorded timestamp, to keep related things together.

File: akka-stream/src/main/scala/akka/stream/impl/Throttle.scala [hunk 6]
```diff
@@ -42,26 +68,34 @@ private[stream] class Throttle[T](cost: Int,
...
                 currentElement = Some(elem)
-                val waitTime = (elementCost - currentTokens) / speed
-                previousTime = currentTime + waitTime
-                scheduleOnce(timerName, waitTime.nanos)
+                val waitNanos = (elementCost - refreshedTokens) * nanosBetweenTokens
+                // due to the rounding down in tokensArrived this will track the idealized time when the last
+                // recorded integer token arrived, which in this case will be the time once we have all the
+                // tokens available to emit this element (i.e. this is calculated *ahead* as this arrival time
+                // is currently in the future and will be valid after the timer fires).
+                arrivalOfLastRecordedToken += elementCost * nanosBetweenTokens
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: This may still lie in the past if refreshedTokens was capped and only therefore did not contain enough for this element.

File: akka-stream/src/main/scala/akka/stream/impl/Throttle.scala (hunk 4 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: Well, refreshedTokens is capped, but that is wrong since I wanted the delta here, not the final. I need the delta and it will be correct.

File: akka-stream/src/main/scala/akka/stream/impl/Throttle.scala (hunk 5 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: ok

File: akka-stream/src/main/scala/akka/stream/impl/Throttle.scala (hunk 4 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: (i.e. it is incorrect, but for a different reason)

File: akka-stream/src/main/scala/akka/stream/impl/Throttle.scala (hunk 3 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: It does not matter where it is done since the bucket is full already, therefore I’d prefer to keep the initialization at the declaration site.

File: akka-stream/src/main/scala/akka/stream/impl/Throttle.scala (hunk 6 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: the issue here is that the capping should not happen there. I.e. if there has been enough time elapsed that all elements were consumed from the token *and* additional time has elapsed, then we are still fine. I misplaced the capping here.

**Comment by drewhk** [comment]
Ok, so I think it is better to factor out the token bucket itself and add more fine-grained tests. It will also make it reusable in other contexts.

File: akka-stream/src/main/scala/akka/stream/impl/Throttle.scala (hunk 3 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: It does slightly matter, since preStart is closer to actual startup. It is not that important though.

**Comment by drewhk** [comment]
Anyway, I updated the current one. I will then factor the implementation and add more tests.

**Comment by drewhk** [comment]
Well, the moment when you exit the building, go too far to turn back, and you realize that you messed up again :) Anyway, this needs a better test suite that covers all the edge cases.

**Comment by drewhk** [comment]
Now with a fresh new token bucket :)

File: akka-actor/src/main/scala/akka/util/TokenBucket.scala [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+   *
+   * If a non-zero value is returned, it is the responsibility of the caller to not call this method before the
+   * returned delay has been elapsed (but can be called later). This class does not check or protect against early
+   * calls.
+   *
+   * @param cost How many tokens the element costs. Can be larger than the capacity of the bucket.
+   * @return
+   */
+  def offer(cost: Long): Long = {
+    require(cost >= 0, "Cost must be non-negative")
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/util/TokenBucket.scala]: how is it now, does this allocate something?

Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/TokenBucket.scala]: Allocates function I guess:

```
def require(requirement: Boolean, message: => Any) 
```

Comment (drewhk) [inline akka-actor/src/main/scala/akka/util/TokenBucket.scala]: meh, replacing with an `if`

**Comment by patriknw** [comment]
LGTM

File: akka-stream/src/main/scala/akka/stream/impl/Throttle.scala [hunk 8]
```diff
@@ -20,70 +21,62 @@ private[stream] class Throttle[T](cost: Int,
...
+  require(per.toNanos > 0, "per time must be > 0")
   require(!(mode == ThrottleMode.Enforcing && maximumBurst < 0), "maximumBurst must be > 0 in Enforcing mode")
+  require(per.toNanos >= cost, "Rates larger than 1 unit / nanosecond are not supported")
+
+  // There is some loss of precision here because of rounding, but this only happens if nanosBetweenTokens is very
+  // small which is usually at rates where that precision is highly unlikely anyway as the overhead of this stage
+  // is likely higher than the required accuracy interval.
+  private val nanosBetweenTokens = per.toNanos / cost
+  private val timerName: String = "ThrottleTimer"
+  private val tokenBucket = new DefaultTokenBucket(maximumBurst, nanosBetweenTokens)
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: oh no, this must go into the logic...

**Comment by drewhk** [comment]
I will rebase https://github.com/akka/akka/pull/20022 on top of this

File: akka-actor-tests/src/test/scala/akka/util/TokenBucketSpec.scala [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+
+import scala.util.Random
+
+class TokenBucketSpec extends AkkaSpec {
+
+  class TestBucket(_cap: Long, _period: Long) extends TokenBucket(_cap, _period) {
+    var currentTime: Long = 0L
+  }
+
+  "A Token Bucket" must {
```
Comment (viktorklang) [inline akka-actor-tests/src/test/scala/akka/util/TokenBucketSpec.scala]: Looks like this spec would be suited as a ScalaCheck spec?

File: akka-actor/src/main/scala/akka/util/TokenBucket.scala [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+      delay
+    }
+  }
+
+}
+
+/**
+ * Default implementation of [[TokenBucket]] that uses `System.nanoTime` as the time source.
+ */
+final class DefaultTokenBucket(_cap: Long, _period: Long) extends TokenBucket(_cap, _period) {
```
Comment (viktorklang) [inline akka-actor/src/main/scala/akka/util/TokenBucket.scala]: Why not: NanoTimeTokenBucket? (Default is a role not an implementation :))

File: akka-stream/src/main/scala/akka/stream/impl/Throttle.scala [hunk 11]
```diff
@@ -20,70 +21,63 @@ private[stream] class Throttle[T](cost: Int,
...
-                scheduleOnce(timerName, waitTime.nanos)
-              case Enforcing ⇒ failStage(new RateExceededException("Maximum throttle throughput exceeded"))
-            }
+          if (delayNanos == 0L) push(out, elem)
           else {
-            lastTokens = currentTokens - elementCost
-            previousTime = currentTime
-            push(out, elem)
+            if (enforcing) failStage(new RateExceededException("Maximum throttle throughput exceeded."))
+            currentElement = elem
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: else?

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/Throttle.scala]: ah yes, failStage does not exit the method :)

File: akka-actor-tests/src/test/scala/akka/util/TokenBucketSpec.scala (hunk 9 above)
Comment (drewhk) [inline akka-actor-tests/src/test/scala/akka/util/TokenBucketSpec.scala]: I don't really see how that helps, the cases below all test very specific cases, so if one of them fails you already have a clue where to start (as they are human curated). The stress test in the bottom would be something more ScalaCheck-like (testing a mutable object against a reference implementation), but that kind of test here takes lot of boilerplate in ScalaCheck (at least the last time I checked).

File: akka-actor/src/main/scala/akka/util/TokenBucket.scala (hunk 10 above)
Comment (drewhk) [inline akka-actor/src/main/scala/akka/util/TokenBucket.scala]: good point

File: akka-actor-tests/src/test/scala/akka/util/TokenBucketSpec.scala (hunk 9 above)
Comment (viktorklang) [inline akka-actor-tests/src/test/scala/akka/util/TokenBucketSpec.scala]: It helps because the person who is going to add or fix test cases do not have to derive, manually, the laws governing the implementation, but can rather read them based on the generators. :)

Comment (drewhk) [inline akka-actor-tests/src/test/scala/akka/util/TokenBucketSpec.scala]: The kind of laws that ScalaCheck can express are very limiting. The laws here are expressed in the last test which expresses a behavioral equality (bi-simulation if you like) between the IUT and a reference model. ScalaCheck rather ugly in these kind of tests: https://github.com/rickynils/scalacheck/blob/master/doc/UserGuide.md#stateful-testing

Comment (viktorklang) [inline akka-actor-tests/src/test/scala/akka/util/TokenBucketSpec.scala]: Alright, lets revisit

Comment (drewhk) [inline akka-actor-tests/src/test/scala/akka/util/TokenBucketSpec.scala]: Maybe it is time to write a tool for such cases ourselves? I am yet to figure out how can one be nicely readable and not more boilerplate than rolling your own (most of these tests rely on specific environmental help like dummy clocks, stream probes, etc) but this is something I would be happy to investigate.

**Comment by patriknw** [comment]
LGTM again
squash?

**Comment by drewhk** [comment]
It was deliberately not squashed because one is a bugfix with corresponding test updates and the other is the rewrite. I found it confusing to squash together. WDYT?

**Comment by patriknw** [comment]
sounds good (just checking that it wasn't an oversight), please merge when validated

**Comment by drewhk** [comment]
Fixes #19862
